### PR TITLE
[RHOL-445]: Add convenience methods for String

### PR DESCRIPTION
**Pull request by itegulov**
## Overview
Add convenience methods for String:

- length method
- slice method
- interpolation operator %
- concatenation operator ++

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
[RHOL-445](https://rchain.atlassian.net/browse/RHOL-445)

### Complete this checklist before you submit the PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Is there any mechanism for converting ground terms into Strings (I do not want to override the case class geenrated toString)? Should I use `PrettyPrinter` or maybe implement instances for `cats.Show` typeclass?

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 1]
```diff
@@ -827,6 +894,44 @@ object Reduce {
...
 
+    private[this] def length: MethodType =
+      (p: Par, args: Seq[Par]) =>
+        (env: Env[Par]) => {
+          def length(baseExpr: Expr): M[Expr] =
+            baseExpr.exprInstance match {
+              case GString(string) =>
+                Applicative[M].pure[Expr](GInt(string.length))
+              case _ =>
+                s.raiseError(ReduceError("Error: method 'length' can only be called on Strings."))
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: According to latest comment/discussions we should use errors messages like:
> Error: Method `length` is not defined on _the_type_

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 2]
```diff
@@ -827,6 +894,44 @@ object Reduce {
...
+
+    private[this] def slice: MethodType =
+      (p: Par, args: Seq[Par]) =>
+        (env: Env[Par]) => {
+          def slice(baseExpr: Expr, from: Int, until: Int): M[Par] =
+            baseExpr.exprInstance match {
+              case GString(string) =>
+                Applicative[M].pure[Par](GString(string.slice(from, until)))
+              case _ =>
+                s.raiseError(ReduceError("Error: method 'slice' can only be called on Strings."))
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Same comment applies here.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 3]
```diff
@@ -537,7 +537,74 @@ object Reduce {
...
+                           }
+                       case (_: GString, _) =>
+                         s.raiseError(ReduceError("Error: String can be interpolated only by Map"))
+                       case _ =>
+                         s.raiseError(ReduceError("Error: only Strings can be interpolated"))
+                     }
+          } yield result
+        case EPlusPlusBody(EPlusPlus(p1, p2)) =>
+          for {
+            v1 <- evalSingleExpr(p1.get)
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: I think you can drop `.get` . There are some residues here and there but thanks to 
```
Par p1 = 1 [(scalapb.field).no_box = true];
```

in proto files it won't wrap in `Option`.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 4]
```diff
@@ -537,7 +537,74 @@ object Reduce {
...
+            v2 <- evalSingleExpr(p2.get)
+            result <- (v1.exprInstance, v2.exprInstance) match {
+                       case (GString(lhs), GString(rhs)) =>
+                         Applicative[M].pure[Expr](GString(lhs + rhs))
+                       case _ =>
+                         s.raiseError(ReduceError("Error: only Strings can be concatenated"))
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Same comment about error messages applies.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 5]
```diff
@@ -537,7 +537,74 @@ object Reduce {
...
+                  ReduceError(s"Error: interpolation doesn't support ${value.getClass}")
+                )
+              case _ =>
+                s.raiseError[(String, String)](
+                  ReduceError("Error: interpolation Map should only contain String keys")
+                )
+            }
+          for {
+            v1 <- evalSingleExpr(p1.get)
+            v2 <- evalSingleExpr(p2.get)
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Drop `.get`s .

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 6]
```diff
@@ -537,7 +537,74 @@ object Reduce {
...
+                                 case (k, v) => current.startsWith("${" + k + "}")
+                               } match {
+                                 case Some((k, v)) =>
+                                   result ++= v
+                                   current = current.drop(k.length + 3)
+                                 case None =>
+                                   result += current.head
+                                   current = current.tail
+                               }
+                             }
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Couldn't we just `replace` elements of format `${x}` in the string? 
https://scalafiddle.io/sf/nQz5nW1/0

Comment (itegulov) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: I also thought so (my second commit implementation looked just like yours), but unfortunately it is not so easy. Consider the following situation:
"Hello, ${name1}! Bye, ${name2}!" % {"name1": "Bob ${name2}", "name2": "Alice"}
Your implementation returns "Hello, Bob Alice! Bye, Alice!", while mine returns "Hello, Bob ${name2}! Bye, Alice!"

Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Ok, makes sense. Can you interpolating part to the method?

Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Also, I would prefer to see this implemented as a regex but it's just my personal taste. We can keep it as is.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 3 above)
Comment (itegulov) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Are you sure? Every single case here has them. Should I also delete them as well?

Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: If it compiles after removing then it will work.

**Comment by itegulov** [comment]
Still working on the tests for normalization and reduction

**Comment by itegulov** [comment]
Wrong target

**Pull request closed without merging**